MHArmory.Search/OpenCL/SearchDataSerializer.cs

```diff
@@ -0,0 +1,243 @@
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using MHArmory.Core.DataStructures;
+
+namespace MHArmory.Search.OpenCL
+{
+    public class SearchDataSerializer
+    {
+        public SerializedSearchParameters Serialize(ISolverData data)
+        {
+            var parameters = new SerializedSearchParameters();
+            
+            var heads = data.AllHeads.Where(x => x.IsSelected).Select(x => x.Equipment).ToList();
+            var chests = data.AllChests.Where(x => x.IsSelected).Select(x => x.Equipment).ToList();
+            var gloves = data.AllGloves.Where(x => x.IsSelected).Select(x => x.Equipment).ToList();
+            var waists = data.AllWaists.Where(x => x.IsSelected).Select(x => x.Equipment).ToList();
+            var legs = data.AllLegs.Where(x => x.IsSelected).Select(x => x.Equipment).ToList();
+            var charms = data.AllCharms.Where(x => x.IsSelected).Select(x => x.Equipment).ToList();
+
+            var allArmorPieces = heads.Concat(chests).Concat(gloves).Concat(waists).Concat(legs).Cast<IArmorPiece>().ToList();
+            var allEquipment = allArmorPieces.Concat(charms).ToList();
+
+            var armorSetParts = allArmorPieces.Where(x => x.ArmorSetSkills != null).SelectMany(x => x.ArmorSetSkills.SelectMany(y => y.Parts)).ToList();
+            var desiredAbilitiesSkillSet = new HashSet<int>(data.DesiredAbilities.Select(x => x.Skill.Id));
+            IEnumerable<IArmorSetSkillPart> desiredArmorSetParts = armorSetParts.Where(x => x.GrantedSkills.Any(y => desiredAbilitiesSkillSet.Contains(y.Skill.Id)));
+            IEnumerable<int> desiredSkillIDs = data.DesiredAbilities.Select(x => x.Skill.Id);
+
+            var maps = new SearchIDMaps();
+
+            maps.EquipmentIdMap = CreateIDMap(allEquipment.Select(x => new Tuple<int, EquipmentType>(x.Id, x.Type)));
```

Comment:
I remember I told you to use a tuple, but actually I recommend you to use the `KeyValuePair<K, V>` type, or the native tuple syntax, such as:

```CS
maps.EquipmentIdMap = CreateIDMap(allEquipment.Select(x => (x.Id, x.Type)));
```

This should still work fine because native tuple are structures very friend for `Dictionary<K, V>` jeys, will still work in method `CreateIDMap`.